School info schools consitency

drewsamnick opened the pull request:
I discovered several cases where school_infos referenced schools but had conflicting data and also found multiple rows for the same school in school_infos. This fixes these problems from growing in the future by:
1. Making sure new rows inserted into school_infos match what is in schools
2. updating the reduplication logic to match based just on the school_id

I did not fail creation of new school infos when conflicting data was passed in since I don't know where or when that happens and I don't want to accidentally prevent somebody from moving on. Similarly, I only fill in the data from the School on creation so we don't muck with existing data unintentionally.

aoby commented on dashboard/app/models/school_info.rb:
```diff
@@ -73,6 +74,51 @@ def school_state=(input)
+      original[:school_other] = school_other
+      original[:school_name] = school_name
+      original[:full_address] = full_address
+      original[:validation_type] = validation_type
```
This can be 
```ruby
original = attributes.slice *%w(
  country
  school_type
  state
)
```
etc with all the fields

drewsamnick replied:
I considered something like that but it was about the same amount of code and this way  felt more readable to me.

mirlew commented:
@bencodeorg do you want to review this?

bencodeorg commented:
Added @joannepolsky to review if she has time on her last day, as she spent a lot of time deep in this code recently.

aoby commented on dashboard/app/models/school_info.rb:
```diff
@@ -73,6 +74,51 @@ def school_state=(input)
+
+      # Report if we are overriding a non-nil value that was originally passed in
+      something_overwritten = original.map {|key, value| value && value != self[key]}.reduce {|acc, b| acc || b}
+      if something_overwritten
```
If you use the above approach, and store the keys in an array, then you should just be able to compare hashes here:

```ruby
# from above
relevant_attributes = %w(
  country
  school_type
) # etc

# here:
unless original == attributes.slice(*relevant_attributes)
  # log to honeybadger
```

drewsamnick replied:
I don't think that works since I don't want to compare nils from original with anything in the final set of attributes.

drewsamnick commented:
I added tests for the Honeybadger notifications. I also modified the existing tests so they don't use inconsistent school data by default since this was breaking some other tests.
